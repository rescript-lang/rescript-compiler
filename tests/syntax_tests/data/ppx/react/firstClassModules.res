@@jsxConfig({version: 4, mode: "classic"})

module Select = {
  module type T = {
    type key
    type t
  }

  @react.component
  let make = (
    type a key,
    ~model as module(T: T with type t = a and type key = key),
    ~selected: option<key>,
    ~onChange: option<key> => unit,
    ~items: array<a>,
  ) => {
    let _ = (model, selected, onChange, items)
    <div />
  }
}

module C6 = {
  module type Comp = {
    let xx: int
    @react.component
    let make: unit => React.element
  }

  @react.component
  let make = (~comp as module(Comp: Comp), ~x as (a, b)) => Comp.xx
}

module External = {
  module type T = {
    type key
    type t
  }

  @react.component @module("c")
  external make: (
    ~model: module(T with type t = 'a and type key = 'key),
    ~selected: option<'key>,
    ~onChange: option<'key> => unit,
    ~items: array<'a>,
  ) => React.element = "default"
}
