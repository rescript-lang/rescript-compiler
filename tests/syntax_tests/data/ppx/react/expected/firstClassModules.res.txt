@@jsxConfig({version: 4, mode: "classic"})

module Select = {
  module type T = {
    type key
    type t
  }
  @res.jsxComponentProps
  type props<'model, 'selected, 'onChange, 'items> = {
    model: 'model,
    selected: 'selected,
    onChange: 'onChange,
    items: 'items,
  }

  let make = (type a key, props: props<_, option<key>, option<key> => unit, array<a>>) => {
    let {model: module(T: T with type t = a and type key = key)} = props
    let selected: option<key> = props.selected
    let onChange: option<key> => unit = props.onChange
    let items: array<a> = props.items

    let _ = (model, selected, onChange, items)
    ReactDOM.createDOMElementVariadic("div", [])
  }
  let make = {
    let \"FirstClassModules$Select" = (props: props<_>) => make(props)

    \"FirstClassModules$Select"
  }
}

module C6 = {
  module type Comp = {
    let xx: int
    @res.jsxComponentProps
    type props = {}

    let make: React.componentLike<props, React.element>
  }
  @res.jsxComponentProps
  type props<'comp, 'x> = {comp: 'comp, x: 'x}

  let make = (props: props<_, _>) => {
    let {comp: module(Comp: Comp)} = props
    let x = props.x
    Comp.xx
  }
  let make = {
    let \"FirstClassModules$C6" = (props: props<_>) => make(props)

    \"FirstClassModules$C6"
  }
}

module External = {
  module type T = {
    type key
    type t
  }
  @res.jsxComponentProps @live
  type props<'model, 'selected, 'onChange, 'items> = {
    model: 'model,
    selected: 'selected,
    onChange: 'onChange,
    items: 'items,
  }

  @module("c")
  external make: React.componentLike<
    props<
      module(T with type t = 'a and type key = 'key),
      option<'key>,
      option<'key> => unit,
      array<'a>,
    >,
    React.element,
  > = "default"
}
